### DISCUSSION TITLE
How to deploy to Netlify or Vercel between a pnpm workspace

### DISCUSSION THREAD

**Discussion by alvarosabu** [nuxt/nuxt#17308] [Questions] [answered]
### Discussed in https://github.com/nuxt/framework/discussions/4943

<div type='discussions-op-text'>

<sup>Originally posted by **alvarosabu** May 12, 2022</sup>
Hi, there community,

I'm working on a **pnpm workspace** monorepo which contains a Nuxt v3.0.0-rc.3 app that I want to deploy to Netlify | Vercel.

I'm facing an issue on build ci on both platforms, "Failed to resolve" multiple packages including `vue` and `ufo` https://app.netlify.com/sites/alvarosabu-next/deploys/627caac87fa81c000876ba1c

Locally the build is successful

I honestly don't know if it's a Nuxt issue, a Netlify|Vercel one, pnpm, of all. I will appreciate some feedback.

Thanks

## Netlify config
```
[build.environment]
  NODE_VERSION = "16"
  NPM_FLAGS = "--version" # prevent Netlify npm install
[build]
  publish = "apps/portfolio/.output/public"
  command = "npx pnpm@6 i --store=node_modules/.pnpm-store --frozen-lockfile && npx pnpm@6 run build:ci"
```

## Related scripts
```
"build:ci": "pnpm build:ui && pnpm run generate:portfolio",
"generate:portfolio": "pnpm run generate --filter @alvarosabu/portfolio",
 "build:ui": "pnpm run build --filter @alvarosabu/ui"
```

## Error log

```
[log] Nuxt CLI v3.0.0-rc.3
8:20:12 AM: [error] Failed to resolve force included dependency: vue
8:20:12 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:12 AM: [error] Failed to resolve import "vue" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/components.plugin.mjs". Does the file exist?
8:20:12 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:12 AM: [error] Failed to resolve import "vue" from "app.vue". Does the file exist?
8:20:12 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:13 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:13 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:13 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:13 AM: [error] Failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs". Does the file exist?
8:20:13 AM: [error] [vite]: Rollup failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs".
8:20:13 AM: This is most likely unintended because it can break your application at runtime.
8:20:13 AM: If you do want to externalize this module explicitly add it to
8:20:13 AM: `build.rollupOptions.external`
8:20:13 AM: [error] [vite]: Rollup failed to resolve import "ufo" from "virtual:nuxt:/opt/build/repo/apps/portfolio/.nuxt/paths.mjs".
8:20:13 AM: This is most likely unintended because it can break your application at runtime.
8:20:13 AM: If you do want to externalize this module explicitly add it to
8:20:13 AM: `build.rollupOptions.external`
8:20:13 AM:   This is most likely unintended because it can break your application at runtime.
8:20:13 AM:   If you do want to externalize this module explicitly add it to
8:20:13 AM:   `build.rollupOptions.external`
8:20:13 AM:   at onRollupWarning (/opt/build/repo/node_modules/.pnpm/vite@2.9.8/node_modules/vite/dist/node/chunks/dep-e1fc1d62.js:41550:19)
8:20:13 AM:   at onwarn (/opt/build/repo/node_modules/.pnpm/vite@2.9.8/node_modules/vite/dist/node/chunks/dep-e1fc1d62.js:41366:13)
8:20:13 AM:   at Object.onwarn (/opt/build/repo/node_modules/.pnpm/rollup@2.72.0/node_modules/rollup/dist/shared/rollup.js:23184:13)
8:20:13 AM:   at ModuleLoader.handleResolveId (/opt/build/repo/node_modules/.pnpm/rollup@2.72.0/node_modules/rollup/dist/shared/rollup.js:22474:26)
8:20:13 AM:   at /opt/build/repo/node_modules/.pnpm/rollup@2.72.0/node_modules/rollup/dist/shared/rollup.js:22435:26
8:20:13 AM: undefined
8:20:13 AM: /opt/build/repo/apps/portfolio:
8:20:13 AM:  ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @alvarosabu/portfolio@1.1.0 generate: `nuxt generate`
8:20:13 AM: Exit status 1
8:20:13 AM:  ELIFECYCLE  Command failed with exit code 1. (https://ntl.fyi/exit-code-1)
8:20:13 AM:  ELIFECYCLE  Command failed with exit code 1. (https://ntl.fyi/exit-code-1)
8:20:13 AM: ​
8:20:13 AM: ────────────────────────────────────────────────────────────────
8:20:13 AM:   "build.command" failed                                        
8:20:13 AM: ────────────────────────────────────────────────────────────────
8:20:13 AM: ​
8:20:13 AM:   Error message
8:20:13 AM:   Command failed with exit code 1: npx pnpm@6 i --store=node_modules/.pnpm-store --frozen-lockfile && npx pnpm@6 run build:ci (https://ntl.fyi/exit-code-1)
8:20:13 AM: ​
8:20:13 AM:   Error location
8:20:13 AM:   In build.command from netlify.toml:
8:20:13 AM:   npx pnpm@6 i --store=node_modules/.pnpm-store --frozen-lockfile && npx pnpm@6 run build:ci
8:20:13 AM: ​
```</div>

**Comment by alvarosabu** [comment]
I tried the suggestion from the log to add the packages to `build.rollupOptions.external` , it's no longer an error but just warnings

```
vite: {
    build: {
      rollupOptions: {
        external: ['vue', 'ufo', '@vueuse/core'],
        output: {
          globals: {
            vue: 'Vue',
            ufo: 'ufo',
            '@vueuse/core': 'VueUseCore',
          },
        },
      },
    },
  },
```

**Comment by alvarosabu** [comment]
Now I'm getting and error when deploying the site in Netlify: `Deploy did not succeed: Deploy directory '.output/public' does not exist` 

```
5:51:21 PM: [info] Client built in 5791ms
5:51:21 PM: [info] Building server...
5:51:24 PM: [success] Server built in 2781ms
5:51:24 PM: [success] [nitro] Generated public dist
5:51:24 PM: [info] [nitro] Initializing prerenderer
5:51:27 PM: [info] [nitro] Prerendering 3 initial routes with crawler
5:51:27 PM: [log] [nitro]   ├─ / (undefinedms) (ReferenceError: Response is not defined)
5:51:27 PM: [log] [nitro]   ├─ /200 (undefinedms) (ReferenceError: Response is not defined)
5:51:27 PM: [log] [nitro]   ├─ /404 (undefinedms) (ReferenceError: Response is not defined)
5:51:27 PM: [success] You can now deploy `.output/public` to any static hosting!
5:51:27 PM: ​
5:51:27 PM: (build.command completed in 1m 34.9s)
5:51:27 PM: ​
5:51:27 PM: ────────────────────────────────────────────────────────────────
5:51:27 PM:   2. Deploy site                                                
5:51:27 PM: ────────────────────────────────────────────────────────────────
5:51:27 PM: ​
5:51:27 PM: ​
5:51:27 PM: ────────────────────────────────────────────────────────────────
5:51:27 PM:   Configuration error                                           
5:51:27 PM: ────────────────────────────────────────────────────────────────
5:51:27 PM: ​
5:51:27 PM:   Error message
5:51:27 PM:   Deploy did not succeed: Deploy directory '.output/public' does not exist
5:51:27 PM: ​
5:51:27 PM:   Resolved config
5:51:27 PM:   build:
5:51:27 PM:     command: npx pnpm@6 i --store=node_modules/.pnpm-store --frozen-lockfile && npx pnpm@6 run build:ci
5:51:27 PM:     commandOrigin: config
5:51:27 PM:     environment:
5:51:27 PM:       - REVIEW_ID
5:51:27 PM:       - NODE_VERSION
5:51:27 PM:       - NPM_FLAGS
5:51:27 PM:     publish: /opt/build/repo/.output/public
5:51:27 PM:     publishOrigin: config
```

Also tried with public directory from monorepo root `apps/portfolio/.output/public` but same result

**Reply by NicoHof** [reply]
> Now I'm getting and error when deploying the site in Netlify: `Deploy did not succeed: Deploy directory '.output/public' does not exist`
> 
> ```
> 5:51:21 PM: [info] Client built in 5791ms
> 5:51:21 PM: [info] Building server...
> 5:51:24 PM: [success] Server built in 2781ms
> 5:51:24 PM: [success] [nitro] Generated public dist
> 5:51:24 PM: [info] [nitro] Initializing prerenderer
> 5:51:27 PM: [info] [nitro] Prerendering 3 initial routes with crawler
> 5:51:27 PM: [log] [nitro]   ├─ / (undefinedms) (ReferenceError: Response is not defined)
> 5:51:27 PM: [log] [nitro]   ├─ /200 (undefinedms) (ReferenceError: Response is not defined)
> 5:51:27 PM: [log] [nitro]   ├─ /404 (undefinedms) (ReferenceError: Response is not defined)
> 5:51:27 PM: [success] You can now deploy `.output/public` to any static hosting!
> 5:51:27 PM: ​
> 5:51:27 PM: (build.command completed in 1m 34.9s)
> 5:51:27 PM: ​
> 5:51:27 PM: ────────────────────────────────────────────────────────────────
> 5:51:27 PM:   2. Deploy site                                                
> 5:51:27 PM: ────────────────────────────────────────────────────────────────
> 5:51:27 PM: ​
> 5:51:27 PM: ​
> 5:51:27 PM: ────────────────────────────────────────────────────────────────
> 5:51:27 PM:   Configuration error                                           
> 5:51:27 PM: ────────────────────────────────────────────────────────────────
> 5:51:27 PM: ​
> 5:51:27 PM:   Error message
> 5:51:27 PM:   Deploy did not succeed: Deploy directory '.output/public' does not exist
> 5:51:27 PM: ​
> 5:51:27 PM:   Resolved config
> 5:51:27 PM:   build:
> 5:51:27 PM:     command: npx pnpm@6 i --store=node_modules/.pnpm-store --frozen-lockfile && npx pnpm@6 run build:ci
> 5:51:27 PM:     commandOrigin: config
> 5:51:27 PM:     environment:
> 5:51:27 PM:       - REVIEW_ID
> 5:51:27 PM:       - NODE_VERSION
> 5:51:27 PM:       - NPM_FLAGS
> 5:51:27 PM:     publish: /opt/build/repo/.output/public
> 5:51:27 PM:     publishOrigin: config
> ```
> 
> Also tried with public directory from monorepo root `apps/portfolio/.output/public` but same result

Had the same issue on Nuxt. Specifing `dist` as deploy folder does the job but `.output/public` fails. I am curious why it is like this? When I do this locally the folder exists and `dist` is just a symlink to it. So why does it not work on Netlify?

**Reply by alvarosabu** [reply]
I wrote a complete tutorial on the hows and whys https://dev.to/alvarosabu/deploying-a-nuxt-site-on-netlify-between-a-pnpm-monorepo-i64 here explains why you need to use `dist`

**Comment by alvarosabu** [comment]
To give an update,  I managed to deploy a basic Nuxt app by using setting `dist` as the publish folder. However, still doesn't work between the pnpm monorepo. The nuxt generate runs but there is no HTML file in the dist folder

**Comment by sibbng** [comment]
I had the same issue with pnpm 7.1.0

Creating `.npmrc` file in root folder and adding this fixed the issue for me:
```
shamefully-hoist=true
```

Or you could do that if you want to be more specific and don't hoist all other dependencies:
```
public-hoist-pattern[]=ufo
public-hoist-pattern[]=vue
```

**Reply by amery** [reply]
could having added `ufo` and `vue` as _devDependency_ break this _"solution"?_

**Reply by amery** [reply]
so, I did remove `ufo`, `vue`, `jiti`, and `typescript` from `devDependencies`, and then added the 4 as `public-hoist-pattern[]=` entries. no success. but with `shamefully-hoist=true` works.

I'm having this problem with just a skeleton project at https://gitlab.com/licensly/licensly/-/tree/shamefully-hoist

I'm wondering if it's related but `nuxt dev` (`target: "static"`) explodes with
```
✔ Nitro built in 299 ms
[nuxt] [request error] Object.defineProperty called on non-object
  at Function.defineProperty (<anonymous>)  
  at $id_a7972da4 (./.nuxt/dist/server/server.mjs:129:8)  
  at __instantiateModule__ (./.nuxt/dist/server/server.mjs:20116:9)  
  at __ssrLoadModule__ (./.nuxt/dist/server/server.mjs:20054:25)  
  at ssrImport (./.nuxt/dist/server/server.mjs:20079:13)  
  at $id_fe0c645f (./.nuxt/dist/server/server.mjs:17:37)  
  at __instantiateModule__ (./.nuxt/dist/server/server.mjs:20116:9)  
  at __ssrLoadModule__ (./.nuxt/dist/server/server.mjs:20054:25)  
  at ./.nuxt/dist/server/server.mjs:20131:22  
  at ModuleJob.run (node:internal/modules/esm/module_job:198:25)
```
server.mjs:129 being
```ts
Object.defineProperty(exports, '__esModule', { value: true });
```

**Comment by alvarosabu** [comment, marked as answer]
With the help of @danielroe I wrote down a tutorial [here](https://dev.to/alvarosabu/deploying-a-nuxt-site-on-netlify-between-a-pnpm-monorepo-i64) on How to successfully deploy a Nuxt Site to Netlify between a pnpm monorepo

All the troubleshooting is there. I hope it helps
